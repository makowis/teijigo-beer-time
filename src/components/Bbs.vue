<template>
  <div class="bbs">
    <h1>一言掲示板</h1>
    <p>
      足跡や、あいさつ、キリ番報告、等に使用していただいて構いません。
      <br>
      誹謗中傷、個人情報などの書き込みはご遠慮ください。
    </p>
    <div class="message-form">
      <div class="message-form-group">
        <label for="nameInput">HN(ハンドルネーム)</label>
        <input type="text" id="nameInput" maxlength="20" v-model="name">
      </div>
      <div class="message-form-group">
        <label for="messageInput">メッセージ</label>
        <input type="text" id="messageInput" maxlength="100" v-model="message">
      </div>
      <button type="button" @click="sendMessage">送信</button>
    </div>
    <div>
      <ul class="message-list">
        <li v-for="(item, key) in messageList" v-bind:key="key">
          <p>
            {{item.message}} by {{item.name}}
          </p>
          <p class="time-label">
            {{item.createdAt}}
          </p>
        </li>
      </ul>
    </div>
  </div>
</template>

<script lang="ts">
import moment from 'moment';
import Vue from 'vue';
import database from '../firebase-config.ts';

interface Message {
  name: string;
  message: string;
  createdAt: Date;
  sortKey: number;
}

interface BbsData {
  messageList: Message[];
  name: string;
  message: string;
}

export default Vue.extend({
  name: 'bbs',
  data(): BbsData {
    return {
      messageList: [],
      name: '',
      message: '',
    };
  },
  created() {
    this.listen();
  },
  methods: {
    listen() {
      database.ref('messages/').on('value', (snapshot) => {
        if (snapshot) {
          const list = snapshot.val();
          const keys = Object.keys(list);

          const values = keys.map((v) => list[v]);
          this.messageList = values.sort((a, b) => {
            if (a.sortKey > b.sortKey) return 1;
            if (a.sortKey < b.sortKey) return -1;
            return 0;
          });
        }
      });
    },
    sendMessage() {
      if (!this.name || !this.message) return;

      database.ref('messages/').push({
        name: this.name,
        message: this.message,
        createdAt: moment(new Date()).format('YYYY/MM/DD H:mm:ss'),
        sortKey: -new Date(),
      });

      this.name = '';
      this.message = '';
    },
  },
});
</script>

<style scoped>
.message-form {
  padding: 10px;
  box-sizing: border-box;
}

@media all and (min-width: 480px) {
  .message-form {
    width: 400px;
    margin: auto;
  }
}

.message-form-group {
  margin-bottom: 10px;
}

.message-form-group label {
  font-weight: bold;
}

.message-form-group input {
  display: block;
  box-sizing: border-box;
  padding: 5px;
  width: 100%;
}

.message-form button {
  background-color: gray;
  width: 100px;
  color: white;
  padding: 5px;
  margin-right: 10px;
  margin-left: 10px;
}

.message-list {
  padding: 0;
  margin-left: auto;
  margin-right: auto;
  list-style: none;
  text-align: left;
  max-width: 90%;
}

.message-list li {
  position: relative;
  background-color: white;
  padding: 5px 10px;
  margin-bottom: 5px;
}

.message-list .time-label {
  font-size: smaller;
  text-align: right;
  margin-bottom: 0;
}
</style>
