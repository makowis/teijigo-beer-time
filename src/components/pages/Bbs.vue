<template>
  <tbt-default>
    <div class="bbs">
      <tbt-h1>一言掲示板</tbt-h1>
      <p>
        足跡や、あいさつ、キリ番報告、等に使用していただいて構いません。
        <br />
        誹謗中傷、個人情報などの書き込みはご遠慮ください。
      </p>
      <div class="message-form">
        <div class="message-form-group">
          <label for="nameInput">HN(ハンドルネーム)</label>
          <input type="text" id="nameInput" maxlength="20" v-model="name" />
        </div>
        <div class="message-form-group">
          <label for="messageInput">メッセージ</label>
          <input
            type="text"
            id="messageInput"
            maxlength="100"
            v-model="message"
          />
        </div>
        <button type="button" @click="sendMessage">送信</button>
      </div>
      <div>
        <ul class="message-list">
          <li v-for="(item, key) in messageList" v-bind:key="key">
            <p>{{ item.message }} by {{ item.name }}</p>
            <p class="time-label">
              {{ item.createdAt }}
            </p>
          </li>
        </ul>
      </div>
    </div>
  </tbt-default>
</template>

<script lang="ts">
import moment from 'moment';
import Vue from 'vue';
import Component from 'vue-class-component';
import TbtDefault from '@/components/templates/TbtDefault.vue';
import TbtH1 from '@/components/atoms/TbtH1.vue';
import database from '@/firebase-config';

interface Message {
  name: string;
  message: string;
  createdAt: string;
  sortKey: number;
}

interface BbsData {
  messageList: Message[];
  name: string;
  message: string;
}

@Component({
  components: {
    TbtDefault,
    TbtH1,
  },
})
export default class Bbs extends Vue implements BbsData {
  messageList: Message[] = [];

  name: string = '';

  message: string = '';

  created() {
    this.listen();
  }

  listen() {
    database
      .ref('messages/')
      .on('value', (snapshot: firebase.database.DataSnapshot | null) => {
        if (snapshot) {
          const list = snapshot.val();
          const keys = Object.keys(list);

          const values = keys.map((v) => list[v]);
          this.messageList = values.sort((a: Message, b: Message) => {
            if (a.sortKey > b.sortKey) return 1;
            if (a.sortKey < b.sortKey) return -1;
            return 0;
          });
        }
      });
  }

  sendMessage() {
    if (!this.name || !this.message) return;

    const message: Message = {
      name: this.name,
      message: this.message,
      createdAt: moment(new Date()).format('YYYY/MM/DD H:mm:ss'),
      sortKey: -new Date(),
    };
    database.ref('messages/').push(message);

    this.name = '';
    this.message = '';
  }
}
</script>

<style scoped>
.message-form {
  padding: 10px;
  box-sizing: border-box;
}

@media all and (min-width: 480px) {
  .message-form {
    width: 400px;
    margin: auto;
  }
}

.message-form-group {
  margin-bottom: 10px;
}

.message-form-group label {
  font-weight: bold;
}

.message-form-group input {
  display: block;
  box-sizing: border-box;
  padding: 5px;
  width: 100%;
}

.message-form button {
  background-color: gray;
  width: 100px;
  color: white;
  padding: 5px;
  margin-right: 10px;
  margin-left: 10px;
}

.message-list {
  padding: 0;
  margin-left: auto;
  margin-right: auto;
  list-style: none;
  text-align: left;
  max-width: 90%;
}

.message-list li {
  position: relative;
  background-color: white;
  padding: 5px 10px;
  margin-bottom: 5px;
}

.message-list .time-label {
  font-size: smaller;
  text-align: right;
  margin-bottom: 0;
}
</style>
