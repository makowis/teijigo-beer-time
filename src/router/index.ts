import Vue from 'vue';
import Router from 'vue-router';
import Top from '@/components/pages/Top.vue';
import Disco from '@/components/pages/Disco.vue';
import Bbs from '@/components/pages/Bbs.vue';
import Member from '@/components/pages/Member.vue';
import Beenya from '@/components/pages/Beenya.vue';
import NotFound from '@/components/pages/NotFound.vue';

Vue.use(Router);

export default new Router({
  mode: 'history',
  routes: [
    {
      path: '/',
      name: 'Top',
      component: Top,
    },
    {
      path: '/disco',
      name: 'Disco',
      component: Disco,
    },
    {
      path: '/bbs',
      name: 'Bbs',
      component: Bbs,
    },
    {
      path: '/member',
      name: 'Member',
      component: Member,
    },
    {
      path: '/beenya',
      name: 'Beenya',
      component: Beenya,
    },
    {
      path: '*',
      name: 'NotFound',
      component: NotFound,
    },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (savedPosition) {
      return savedPosition;
    }
    return { x: 0, y: 0 };
  },
});
