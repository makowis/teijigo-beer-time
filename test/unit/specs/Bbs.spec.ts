import Vue from 'vue';
import Bbs from '@/components/pages/Bbs.vue';
import router from '@/router/index';

describe('Bbs.vue', () => {
  const Constructor = Vue.extend(Bbs);
  const vm = new Constructor({ router }).$mount();

  it('should match the snapshot', () => {
    expect(vm.$el).toMatchSnapshot();
  });
});
