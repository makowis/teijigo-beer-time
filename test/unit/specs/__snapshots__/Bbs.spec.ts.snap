// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bbs.vue should match the snapshot 1`] = `
<div
  id="app"
>
  <nav
    class="menuContainer"
    id="menu-container"
  >
    <ul
      class="menu"
    >
      <li
        class="menuItem"
      >
        <a
          class="router-link-exact-active router-link-active"
          href="/"
        >
           トップ 
        </a>
      </li>
       
      <li
        class="menuItem"
      >
        <a
          class=""
          href="/disco"
        >
           DISCOGRAPHY 
        </a>
      </li>
       
      <li
        class="menuItem"
      >
        <a
          class=""
          href="/member"
        >
           メンバー紹介 
        </a>
      </li>
       
      <li
        class="menuItem"
      >
        <a
          class=""
          href="/bbs"
        >
           一言掲示板 
        </a>
      </li>
       
      <li
        class="menuItem"
      >
        <a
          class=""
          href="/beenya"
        >
           びーにゃの部屋 
        </a>
      </li>
       
      <li
        class="menuItem"
      >
        <a
          href="https://teijigo.booth.pm/items/1093403"
          target="_blank"
        >
          
        通販(BOOTH)
      
        </a>
      </li>
    </ul>
  </nav>
   
  <div
    class="mainContainer"
    id="main-container"
  >
    <a
      class="router-link-exact-active router-link-active"
      href="/"
    >
      <img
        class="logo"
        src="@/assets/logo_1.png"
      />
    </a>
     
    <div
      class="bbs"
    >
      <h1
        class="h1"
      >
        一言掲示板
      </h1>
       
      <p>
        
      足跡や、あいさつ、キリ番報告、等に使用していただいて構いません。
      
        <br />
        
      誹謗中傷、個人情報などの書き込みはご遠慮ください。
    
      </p>
       
      <div
        class="message-form"
      >
        <div
          class="message-form-group"
        >
          <label
            for="nameInput"
          >
            HN(ハンドルネーム)
          </label>
           
          <input
            id="nameInput"
            maxlength="20"
            type="text"
          />
        </div>
         
        <div
          class="message-form-group"
        >
          <label
            for="messageInput"
          >
            メッセージ
          </label>
           
          <input
            id="messageInput"
            maxlength="100"
            type="text"
          />
        </div>
         
        <button
          type="button"
        >
          送信
        </button>
      </div>
       
      <div>
        <ul
          class="message-list"
        />
      </div>
    </div>
     
    <footer
      class="footer"
    >
      <p>
        ©定時後ビールタイム
      </p>
    </footer>
  </div>
</div>
`;
